src/stan/callbacks/concurrent_writer.hpp

```diff
@@ -0,0 +1,212 @@
+#ifndef STAN_CALLBACKS_CONCURRENT_WRITER_HPP
+#define STAN_CALLBACKS_CONCURRENT_WRITER_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <tbb/concurrent_queue.h>
+#include <condition_variable>
+#include <functional>
+#include <string>
+#include <thread>
+#include <vector>
+
+namespace stan::callbacks {
+#ifdef STAN_THREADS
+/**
+ * Takes a writer and makes it thread safe via multiple queues.
+ * On construction, a thread is spawned to write to the writer.
+ * This class uses an `std::thread` instead of a tbb task graph because
+ * of deadlocking issues. A deadlock can happen in two major cases.
+ * 1. If TBB gives all threads a task, and all threads hit an instance of max
+ * capacity. TBB can choose to wait for a thread to finish instead of spinning
+ * up the write thread. So to circumvent that issue, we use an std::thread.
+ * 2. If the bounded queues are full but the consumer thread is not scheduled
+ *  because there are more busy threads than the number of threads available.
+ * The producer threads are blocked because the queues are full. The consumer
+ * thread is blocked because the producer thread is spinning. Then we have a
+ * deadlock because the consumer thread is blocked because the producer threads
+ * are blocked.
+ * i.e. queue(full)->producer(blocked)->consumer(blocked)->producer(blocked)
+ * To circumvent this issue, we check in the producer threads if
+ * the queues are almost* full and if they are, we make a lock and wait for
+ * the consumer thread to signal it's queues are not longer at capacity. This
+ * frees a thread for the consumer thread to write to the writer. Once the
+ * consumer thread is finished writing, it will notify all the producer threads
+ * to continue sending data. The check for the queues being almost full is
+ * done by checking if the size of the queue is greater than the max capacity
+ * minus the number of threads times 2. This is a heuristic to make sure that,
+ * if some threads slip past the check and write to the queue, the bounded queue
+ * will still not be full.
+ *
+ * @tparam Writer A type that inherits from `writer`
+ */
+template <typename Writer>
+struct concurrent_writer {
+  // A reference to the writer to write to
+  std::reference_wrapper<Writer> writer;
+  // Number of null writes queued
+  std::atomic<int> null_writes_queued{0};
+  // Queue for string messages
+  tbb::concurrent_bounded_queue<std::string> str_messages_{};
+  // Queue for vector of strings messages
+  tbb::concurrent_bounded_queue<std::vector<std::string>> vec_str_messages_{};
```

Comment:
Thinking about this more -- any non-numeric content would be erroneous to write in a concurrent context, and should be written to the underlying writer to ensure ordering. IMO, we should remove this functionality for clarity